# Exercise 02 -  OLAP Cubes - Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](https://video.udacity-data.com/topher/2021/August/61120d38_pagila-star/pagila-star.png)

Start by creating and connecting to the database by running the cells below.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

createdb: database creation failed: ERROR:  database "pagila_star" already exists


### Connect to the local database where Pagila is loaded

In [2]:
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila


'Connected: student@pagila'

### Star Schema

<img src="pagila-star.png" width="50%"/>

In [3]:
%%sql
DROP TABLE IF EXISTS factSales;
DROP TABLE IF EXISTS dimDate;
DROP TABLE IF EXISTS dimCustomer;
DROP TABLE IF EXISTS dimStore;
DROP TABLE IF EXISTS dimMovie;

 * postgresql://student:***@127.0.0.1:5432/pagila
Done.
Done.
Done.
Done.
Done.


[]

In [4]:
%%time
%%sql

CREATE TABLE dimDate
(
  date_key integer NOT NULL PRIMARY KEY,
  date date NOT NULL,
  year smallint NOT NULL,
  quarter smallint NOT NULL,
  month smallint NOT NULL,
  day smallint NOT NULL,
  week smallint NOT NULL,
  is_weekend boolean
);

CREATE TABLE dimCustomer
(
  customer_key SERIAL PRIMARY KEY,
  customer_id  smallint NOT NULL,
  first_name   varchar(45) NOT NULL,
  last_name    varchar(45) NOT NULL,
  email        varchar(50),
  address      varchar(50) NOT NULL,
  address2     varchar(50),
  district     varchar(20) NOT NULL,
  city         varchar(50) NOT NULL,
  country      varchar(50) NOT NULL,
  postal_code  varchar(10),
  phone        varchar(20) NOT NULL,
  active       smallint NOT NULL,
  create_date  timestamp NOT NULL,
  start_date   date NOT NULL,
  end_date     date NOT NULL
);

CREATE TABLE dimMovie
(
  movie_key          SERIAL PRIMARY KEY,
  film_id            smallint NOT NULL,
  title              varchar(255) NOT NULL,
  description        text,
  release_year       integer,
  language           varchar(20) NOT NULL,
  original_language  varchar(20),
  rental_duration    smallint NOT NULL,
  length             smallint NOT NULL,
  rating             varchar(5) NOT NULL,
  special_features   varchar(60) NOT NULL
);
CREATE TABLE dimStore
(
  store_key           SERIAL PRIMARY KEY,
  store_id            smallint NOT NULL,
  address             varchar(50) NOT NULL,
  address2            varchar(50),
  district            varchar(20) NOT NULL,
  city                varchar(50) NOT NULL,
  country             varchar(50) NOT NULL,
  postal_code         varchar(10),
  manager_first_name  varchar(45) NOT NULL,
  manager_last_name   varchar(45) NOT NULL,
  start_date          date NOT NULL,
  end_date            date NOT NULL
);
CREATE TABLE factSales
(
  sales_key        SERIAL PRIMARY KEY,
  date_key         INT NOT NULL REFERENCES dimDate(date_key),
  customer_key     INT NOT NULL REFERENCES dimCustomer(customer_key),
  movie_key        INT NOT NULL REFERENCES dimMovie(movie_key),
  store_key        INT NOT NULL REFERENCES dimStore(store_key),
  sales_amount     decimal(5,2) NOT NULL
);

 * postgresql://student:***@127.0.0.1:5432/pagila
Done.
Done.
Done.
Done.
Done.
CPU times: user 15.2 ms, sys: 2.07 ms, total: 17.3 ms
Wall time: 43.1 ms


[]

In [5]:
%%sql
INSERT INTO dimDate (date_key, date, year, quarter, month, day, week, is_weekend)
SELECT DISTINCT(TO_CHAR(payment_date :: DATE, 'yyyyMMDD')::integer) AS date_key,
       date(payment_date)                                           AS date,
       EXTRACT(year FROM payment_date)                              AS year,
       EXTRACT(quarter FROM payment_date)                           AS quarter,
       EXTRACT(month FROM payment_date)                             AS month,
       EXTRACT(day FROM payment_date)                               AS day,
       EXTRACT(week FROM payment_date)                              AS week,
       CASE WHEN EXTRACT(ISODOW FROM payment_date) IN (6, 7) THEN true ELSE false END AS is_weekend
FROM payment;

INSERT INTO dimCustomer (customer_key, customer_id, first_name, last_name, email, address, address2, district, city, country, postal_code, phone, active, create_date, start_date, end_date)
SELECT c.customer_id AS customer_key,
       c.customer_id,
       c.first_name,
       c.last_name,
       c.email,
       a.address,
       a.address2,
       a.district,
       ci.city,
       co.country,
       a.postal_code,
       a.phone,
       c.active,
       c.create_date,
       now()         AS start_date,
       now()         AS end_date
FROM customer c
JOIN address a  ON (c.address_id = a.address_id)
JOIN city ci    ON (a.city_id = ci.city_id)
JOIN country co ON (ci.country_id = co.country_id);

INSERT INTO dimMovie (movie_key, film_id, title, description, release_year, language, original_language, rental_duration, length, rating, special_features)
SELECT f.film_id      AS movie_key,
       f.film_id,
       f.title,
       f.description,
       f.release_year,
       l.name         AS language,
       orig_lang.name AS original_language,
       f.rental_duration,
       f.length,
       f.rating,
       f.special_features
FROM film f
JOIN language l              ON (f.language_id=l.language_id)
LEFT JOIN language orig_lang ON (f.original_language_id = orig_lang.language_id);

INSERT INTO dimStore (store_key, store_id, address, address2, district, city, country, postal_code, manager_first_name, manager_last_name, start_date, end_date)
SELECT s.store_id    AS store_key,
       s.store_id,
       a.address,
       a.address2,
       a.district,
       c.city,
       co.country,
       a.postal_code,
       st.first_name AS manager_first_name,
       st.last_name  AS manager_last_name,
       now()         AS start_date,
       now()         AS end_date
FROM store s
JOIN staff st   ON (s.manager_staff_id = st.staff_id)
JOIN address a  ON (s.address_id = a.address_id)
JOIN city c     ON (a.city_id = c.city_id)
JOIN country co ON (c.country_id = co.country_id);

INSERT INTO factSales (date_key, customer_key, movie_key, store_key, sales_amount)
SELECT TO_CHAR(p.payment_date :: DATE, 'yyyyMMDD')::integer AS date_key ,
       p.customer_id                                        AS customer_key,
       i.film_id                                            AS movie_key,
       i.store_id                                           AS store_key,
       p.amount                                             AS sales_amount
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id );

 * postgresql://student:***@127.0.0.1:5432/pagila
40 rows affected.
599 rows affected.
1000 rows affected.
2 rows affected.
16049 rows affected.


[]

# Start with a simple cube
TODO: Write a query that calculates the revenue (sales_amount) by day, rating, and city. Remember to join with the appropriate dimension tables to replace the keys with the dimension labels. Sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below.

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

In [6]:
%%time
%%sql

SELECT d.day, m.rating, c.city, SUM(f.sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimCustomer c ON f.customer_key = c.customer_key
JOIN dimMovie m ON f.movie_key = m.movie_key
GROUP BY d.day, m.rating, c.city
ORDER BY revenue DESC
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 3.17 ms, sys: 1.42 ms, total: 4.59 ms
Wall time: 44.7 ms


[(30, 'G', 'San Bernardino', Decimal('24.97')),
 (30, 'NC-17', 'Apeldoorn', Decimal('23.95')),
 (21, 'NC-17', 'Belm', Decimal('22.97')),
 (21, 'G', 'Citt del Vaticano', Decimal('21.97')),
 (30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (28, 'R', 'Mwanza', Decimal('21.97')),
 (1, 'R', 'Qomsheh', Decimal('19.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (22, 'R', 'Yangor', Decimal('19.97')),
 (17, 'G', 'Rajkot', Decimal('19.97')),
 (19, 'PG', 'Najafabad', Decimal('19.96')),
 (30, 'R', 'Fengshan', Decimal('19.95')),
 (21, 'G', 'Wroclaw', Decimal('18.98')),
 (28, 'PG', 'So Leopoldo', Decimal('18.98')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (30, 'G', 'Omdurman', Decimal('18.97')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (19, 'PG', 'Sokoto', Decimal('18.97')),
 (1, 'NC-17', 'Memphis', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95'))]

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

TODO: Write a query that reduces the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below. 

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

In [7]:
%%time
%%sql

SELECT d.day, m.rating, c.city, SUM(f.sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimCustomer c ON f.customer_key = c.customer_key
JOIN dimMovie m ON f.movie_key = m.movie_key
WHERE m.rating = 'PG-13'
GROUP BY d.day, m.rating, c.city
ORDER BY revenue DESC
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 3.69 ms, sys: 212 µs, total: 3.9 ms
Wall time: 18.1 ms


[(30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95')),
 (30, 'PG-13', 'Nagareyama', Decimal('17.98')),
 (21, 'PG-13', 'Parbhani', Decimal('17.98')),
 (20, 'PG-13', 'Baha Blanca', Decimal('17.98')),
 (30, 'PG-13', 'Tanauan', Decimal('17.96')),
 (17, 'PG-13', 'Ikerre', Decimal('17.95')),
 (30, 'PG-13', 'Zhoushan', Decimal('16.98')),
 (30, 'PG-13', 'Newcastle', Decimal('16.97')),
 (30, 'PG-13', 'Yerevan', Decimal('16.97')),
 (8, 'PG-13', 'Pontianak', Decimal('15.98')),
 (30, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (17, 'PG-13', 'Yantai', Decimal('15.98')),
 (20, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (10, 'PG-13', 'Udaipur', Decimal('15.97')),
 (30, 'PG-13', 'Toulouse', Decimal('15.97')),
 (20, 'PG-13', 'Johannesburg', Decimal('15.97'))]

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

TODO: Write a query to create a subcube of the initial cube that includes moves with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

The first few rows of your output should match the table below. 

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>

In [8]:
%%time
%%sql

SELECT d.day, m.rating, c.city, SUM(f.sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimCustomer c ON f.customer_key = c.customer_key
JOIN dimMovie m ON f.movie_key = m.movie_key
WHERE m.rating IN ('PG', 'PG-13')
AND c.city IN ('Bellevue', 'Lancaster')
AND d.day IN (1, 15, 30)
GROUP BY d.day, m.rating, c.city
ORDER BY revenue DESC
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila
6 rows affected.
CPU times: user 3.25 ms, sys: 1.08 ms, total: 4.33 ms
Wall time: 7.87 ms


[(30, 'PG', 'Lancaster', Decimal('12.98')),
 (1, 'PG-13', 'Lancaster', Decimal('5.99')),
 (30, 'PG-13', 'Bellevue', Decimal('3.99')),
 (30, 'PG-13', 'Lancaster', Decimal('2.99')),
 (15, 'PG-13', 'Bellevue', Decimal('1.98')),
 (1, 'PG', 'Bellevue', Decimal('0.99'))]